File: crates/cxx-qt-lib/src/types/qcolor.cpp
// clang-format off
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
// SPDX-FileContributor: Leon Matthes <leon.matthes@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0
#include "cxx-qt-lib/include/qcolor.h"

#include "assertion_utils.h"

// QColor has an enum with six values and a union with the largest being five
// ushorts. This results in (5 * std::uint16) + std::uint32_t = 14, then due to
// compiler padding this results in a sizeof 16 or two pointers.
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/gui/painting/qcolor.h?h=v5.15.6-lts-lgpl#n262
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/gui/painting/qcolor.h?h=v6.2.4#n237
assert_alignment_and_size(QColor, alignof(std::size_t), sizeof(std::size_t[2]));

// QColor still had copy & move constructors in Qt 5 but they were basically
// trivial.
#if (QT_VERSION >= QT_VERSION_CHECK(6, 0, 0))
static_assert(std::is_trivially_copyable<QColor>::value);
#else
static_assert(QTypeInfo<QColor>::isRelocatable);
#endif

static_assert(std::is_trivially_destructible<QColor>::value);

namespace rust {
namespace cxxqtlib1 {

QColor
qcolorInitDefault()
{
  return QColor();
}

QColor
qcolorInitFromRgba(std::int32_t r,
                   std::int32_t g,
                   std::int32_t b,
                   std::int32_t a)
{
  return QColor(r, g, b, a);
}

QColor
qcolorInitFromQColor(const QColor& color)
{
  return QColor(color);
}

}

}

File: tests/qt_types_standalone/rust/src/qcolor.rs
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
// SPDX-FileContributor: Gerhard de Clercq <gerhard.declercq@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use cxx_qt_lib::QColor;

#[cxx::bridge]
mod qcolor_cxx {
    enum ColorTest {
        Rgb_Red,
        Rgb_Green,
        Rgb_Blue,
        Rgb_Transparent,
    }

    unsafe extern "C++" {
        include!("cxx-qt-lib/qcolor.h");
        type QColor = cxx_qt_lib::QColor;
    }

    extern "Rust" {
        fn construct_qcolor(test: ColorTest) -> QColor;
        fn read_qcolor(c: &QColor, test: ColorTest) -> bool;
        fn clone_qcolor(c: &QColor) -> QColor;
    }
}

use qcolor_cxx::ColorTest;

fn construct_qcolor(test: ColorTest) -> QColor {
    match test {
        ColorTest::Rgb_Red => QColor::from_rgb(255, 0, 0),
        ColorTest::Rgb_Green => QColor::from_rgb(0, 255, 0),
        ColorTest::Rgb_Blue => QColor::from_rgb(0, 0, 255),
        ColorTest::Rgb_Transparent => QColor::from_rgba(0, 0, 0, 0),
        _others => panic!("Unsupported test: {}", test.repr),
    }
}

fn read_qcolor(color: &QColor, test: ColorTest) -> bool {
    match test {
        ColorTest::Rgb_Red => {
            color.alpha() == 255 && color.red() == 255 && color.green() == 0 && color.blue() == 0
        }
        ColorTest::Rgb_Green => {
            color.alpha() == 255 && color.red() == 0 && color.green() == 255 && color.blue() == 0
        }
        ColorTest::Rgb_Blue => {
            color.alpha() == 255 && color.red() == 0 && color.green() == 0 && color.blue() == 255
        }
        ColorTest::Rgb_Transparent => {
            color.alpha() == 0 && color.red() == 0 && color.green() == 0 && color.blue() == 0
        }
        _others => panic!("Unsupported test: {}", test.repr),
    }
}

fn clone_qcolor(c: &QColor) -> QColor {
    c.clone()
}
